File: Java/Greedy_Algorithms/minimumCoins.java
package Greedy_Algorithms;

import java.util.Arrays;
import java.util.Collections;
import java.util.Scanner;

public class minimumCoins {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        System.out.println("Enter the amount of money: ");
        int amount = scan.nextInt();

        System.out.println("Enter how many types of coins you have: ");
        int types = scan.nextInt();

        Integer[] coins = new Integer[types];
        System.out.println("Enter the types of coins: ");
        for (int i = 0; i < types; i++) {
            coins[i] = scan.nextInt();
        }

        System.out.println("Minimum number of coins required to obtain the amount: "+minCoins(amount, coins));

    }

    public static int minCoins(int amount, Integer[] coins) {

        //  Sort the coins in descending order
        Arrays.sort(coins, Collections.reverseOrder());
        int min = 0;
        for (Integer coin: coins) {
            if (amount >= coin) {
                int nCoins = (int) Math.floor((double)(amount / coin));
                min = min + nCoins;
                amount -= coin * nCoins;
            }
        }

        return min;
    }
}

File: Golang/Greedy Algorithms/minimumCoins.go
package main

import (
	"fmt"
	"log"
	"math"
	"sort"
)

/*
Consider infinite supply of coins: 10	5	2	1

If someone ask for an amount, how will we give that amount using minimum coins ?
 */

func minCoin(amount int, coins []int) (int, error) {
	if len(coins) == 0 {
		return 0, fmt.Errorf("you didn't specify what coins you have")
	}
	if amount == 0 {
		return 0, fmt.Errorf("you didn't specify an amount")
	}

	var result = 0
	for _, coin := range coins {
		if coin <= amount {
			nCoins := int(math.Floor(float64(amount / coin)))
			result = result + nCoins
			amount -= nCoins * coin
		}
		if amount == 0 {
			break
		}
	}

	return result, nil
}

func main() {
	var amount int
	fmt.Println("Enter the amount: ")
	_, err := fmt.Scanln(&amount)
	if err != nil || amount == 0 {
		log.Fatal(err)
	}

	var types int
	fmt.Println("How many types of coins you have: ")
	_, err = fmt.Scanln(&types)
	if err != nil || types == 0 {
		log.Fatal(err)
	}

	fmt.Println("Enter the types of coins you have: ")
	coins := make([]int, types)
	for i := range coins {
		_, _ = fmt.Scanf("%d", &coins[i])
	}

	// sorting the coins in descending order
	sort.Slice(coins, func(i, j int) bool {
		return coins[i] > coins[j]
	})

	count, err2 := minCoin(amount, coins)
	if err2 != nil {
		log.Fatal(err2)
	}
	fmt.Println("Minimum No of coins required: ", count)

}
